proposed/logging-generator.md

```diff
@@ -0,0 +1,327 @@
+
+    // returns the number of holes
+    public int Count => ... // depends on the type used to store key value pairs
+
+    public static readonly Func<CustomLoggerParamsBuilder, Exception?, string> FormatCallback = (builder, exception) => builder.ToString();
+
+    private CustomLoggerParamsBuilder(int baseLength, int formatHoleCount, LogLevel logLevelEnabled)
+    {
+        // Initialization logic
+        _logLevelEnabled = logLevelEnabled;
+    }
+
+    public bool TryFormat(string s)
+    {
+        // Store and format part as required
+        return true;
+    }
+
+    public bool TryFormat<T>(T t, string s)
+    {
+        // Store and format part as required
+        // Store name hole and value as KeyValuePair<string, object?>
+        return true;
+    }
+
+    public KeyValuePair<string, object?> this[int index]
+    {
+        get => ... // depends on the type used to store key value pairs
+    }
+
+    public IEnumerator<KeyValuePair<string, object>> GetEnumerator()
+    {
+        for (int i = 0; i < Count; ++i)
+        {
+            yield return this[i];
+        }
+    }
+
+    IEnumerator IEnumerable.GetEnumerator()
+    {
+        return GetEnumerator();
+    }
+}
+```
+
+In order to make use of a builder like `CustomLoggerParamsBuilder`, we would need new overloads in the future for LogInformation (and other similar APIs):
+
+```c#
+// We need new overloads to support this call:
+logger.LogInformation($"Welcome to {city}!");
+```
+
+But the sample below shows an application of the builder using an existing logging API:
+```c#
+var city = "Vancouver";
+
+logger.Log(
+    LogLevel.Information,
+    new EventId(1),
+    $"Welcome to {city}!",
+    new Exception(),
+    (s, e) => s);
+```
+The compiler would translate the above call to:
+```c#
+logger.Log<CustomLoggerParamsBuilder>(
+    LogLevel.Information,
+    new EventId(1),
+    builder, // the CustomLoggerParamsBuilder
+    new Exception(),
+    builder.FormatCallback
+);
+```
+Through this new compiler feature, the builder would be able to recognize name holes, (here the city) or even format specifiers. However, currently the proposed builder approach is limited in its ability to get both name holes and format specifiers at once. The logger to is able to identify both of these from message templates and would be good to have this feature available.
+
+Also, an older compiler, the builder overload may not be detectable and therefore the above log APIs could end up calling the wrong overload, which would cause in logging messages to completely lose the structure provided in the input. This could be however be mitigated with an analyzer that warns user when they use string interpolated messages on older compiler versions.
+ 
+### Limitations with string interpolation builder approach:
+
+- With the interpolated string builder approach, there is currently no way to get both name holes and format specifiers, but this is supported today already by the existing logger APIs.
```

Comment:
I'm not sure this is quite true, but it would depend on how creative we want to get with the format string. We could use some currently-unused character to specify a "separator" between the regular format string part and the name you want the hole to have. It would certainly take some design though.

Reply:
@333fred I would love for that. Do we file an issue for this?

Reply:
To be clear, I was not talking about a separate language feature here, just making the proposed builder type above smarter. I would imagine and format changes would be part of this proposal.